util/dvsim/Scheduler.py

```diff
@@ -233,44 +104,372 @@ def on_sigint(signal_received, frame):
         finally:
             signal(SIGINT, old_handler)
 
-        # We got to the end without anything exploding. Return the results for our jobs.
+        # We got to the end without anything exploding. Return the results.
         return self.item_to_status
 
+    def _schedule_items(self, items):
+        '''Convert a heterogenous list of items to run into list of tasks
+        segregated by target and cfg (DUT).'''
 
-class Scheduler:
-    '''An object to run one or more Deploy items'''
+        # Scheduled items are ordered by targets.
+        scheduled = OrderedDict()
 
-    # Max jobs running at one time
-    max_parallel = 16
+        def add_to_scheduled(item):
+            '''Recursively go through the item's dependencies and itself to
+            the task list, associated with target and the cfg name.'''
 
-    # Max jobs dispatched in one go.
-    slot_limit = 20
+            for dep in item.dependencies:
+                add_to_scheduled(dep)
 
-    def __init__(self, items):
-        # An ordered dictionary keyed by target ('build', 'run' or similar).
-        # The value for each target is a TargetScheduler object.
-        self.schedulers = OrderedDict()
+            if item.target not in scheduled:
+                scheduled[item.target] = {}
+
+            if item.sim_cfg not in scheduled[item.target]:
+                scheduled[item.target][item.sim_cfg] = []
+
+            if item not in scheduled[item.target][item.sim_cfg]:
+                scheduled[item.target][item.sim_cfg].append(item)
 
         for item in items:
-            # This works like setdefault, but doesn't construct a TargetScheduler
-            # object unnecessarily.
-            tgt_scheduler = self.schedulers.get(item.target)
-            if tgt_scheduler is None:
-                tgt_scheduler = TargetScheduler(item.target)
-                self.schedulers[item.target] = tgt_scheduler
+            add_to_scheduled(item)
 
-            tgt_scheduler.add_item(item)
+        return scheduled
 
-    def run(self):
-        '''Run all items
+    def _remove_from_scheduled(self, item):
+        '''Helper method to remove an item from the _scheduled list.'''
+
+        if item.sim_cfg in self._scheduled[item.target]:
+            if item in self._scheduled[item.target][item.sim_cfg]:
+                self._scheduled[item.target][item.sim_cfg].remove(item)
+
+            if not self._scheduled[item.target][item.sim_cfg]:
+                del self._scheduled[item.target][item.sim_cfg]
```

Comment:
I think this sort of thing is much easier to read with some local variables to avoid the repeated lookups. Like this, for example:
```python3
        for_tgt = self._scheduled[item.target]
        for_cfg = for_tgt.get(item.sim_cfg)

        if for_cfg is not None:
            for_cfg.remove(item)
            if not for_cfg:
                del for_tgt[item.sim_cfg]
```
It also avoids `del foo[bar][baz]`, which I always have to think about. What do you think?